barretenberg/cpp/pil/vm2/poseidon2_hash.pil

```diff
@@ -4,86 +4,91 @@ include "poseidon2_perm.pil";
     sel = 0;
 
+    // === Input Values ===
+    // Unpadded Length of the inputs to be hashed
+    pol commit input_len;
+    // Compute the IV using the unpadded length. Only used at the start of a new poseidon2 hash
+    pol TWOPOW64 = 18446744073709551616;
+    pol IV = TWOPOW64 * input_len;
+
     // These are the inputs to be hashed this round, we hash chunks of 3 field elements
     pol commit input_0;
     pol commit input_1;
     pol commit input_2;
 
-    // Output of the hash it is matched with the result of the last permutation round
+    // Output of the hash, not permutation round, it is matched with the result of the last permutation round
     pol commit output;
+    // We enforce that the output has to be kept the same per row until we hit the "end" of the computation
+    sel * (1 - LATCH_CONDITION) * (output' - output) = 0;
 
-    pol commit sel;
-    sel * (1 - sel) = 0;
-    sel = execute_perm + end;
-    pol TWOPOW64 = 18446744073709551616;
-
-    pol commit input_len;
-    // Only used at the start of a new poseidon2 hash
-    pol IV = TWOPOW64 * input_len;
-
-
+    // === Control Flow within Subtrace ===
     // Start of a poseidon2 computation
     pol commit start;
     start * (1 - start) = 0;
     // When we end a poseidon, the next row must naturally have a start
-    sel' * (1 - precomputed.first_row) * (start' - end) = 0;
+    sel' * (start' - LATCH_CONDITION) = 0;
 
-    // We track the num of rounds remaining, excluding the first round that has to be performed by the start selector.
-    // We use the padded length to calculate the num of rounds to perform and the unpadded length is used in the IV.
+    // The last row of hashing for the given input
+    pol commit end;
+    end * (1 - end) = 0;
+    // End and first row are NAND - they cannot both be one.
+    end * precomputed.first_row = 0;
+    pol LATCH_CONDITION = end + precomputed.first_row;
+
+    // === Permutation/Round Counting ===
+    // We use the padded length to calculate the num of rounds to perform
     pol commit num_perm_rounds_rem;
     pol commit padding;
-    // Padding can either be 0, 1 or 2
+    // The amount of padding can be 0, 1 or 2.
+    // Note the padded values are not enforced to be zero here, the calling function SHOULD enforce this
     padding * (padding - 1) * (padding - 2) = 0;
     pol PADDED_LEN = input_len + padding;
-    start * ((num_perm_rounds_rem + 1) * 3 - PADDED_LEN) = 0;
-
-
-    // The row with the final result of the poseidon computation
-    pol commit end;
-    end * (1 - end) = 0;
-    // The final result of the output of the hash should match the output from the last permutation (b_0)
-    end * (output - b_0) = 0;
-    pol commit num_perm_rounds_rem_inv;
-    // end == 1 when the num_perm_rounds_rem == 0
-    sel * (num_perm_rounds_rem * (end * (1 - num_perm_rounds_rem_inv) + num_perm_rounds_rem_inv) - 1 + end) = 0;
-
-    // We perform the "squeeze" / perm operation until end
-    pol commit execute_perm;
```

Comment:
We dont need to distinguish this step in reality, we just needed a good definition for `LATCH_CONDITION`